File: src/main/java/com/rojakcoder/archly/Registry.java
package com.rojakcoder.archly;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;

import com.rojakcoder.archly.exceptions.DuplicateEntryException;
import com.rojakcoder.archly.exceptions.EntryNotFoundException;

/**
 * Registry is a generic registry that can be sub-classed for resources and
 * roles.
 */
class Registry {
	static final String DUPLICATE_CHILD = "Entry '%s' is already a child of '%s'";

	static final String DUPLICATE_ENTRIES = "Entry '%s' is already in the registry - cannot add duplicate.";

	static final String NOT_FOUND = "Entry '%s' not in registry.";

	/**
	 * The internal representation of the registry.
	 */
	protected ConcurrentMap<String, String> registry;

	/**
	 * The constructor for creating the registry.
	 * <p>
	 * This is accessible only to child classes for making singleton
	 * implementations.
	 * </p>
	 */
	protected Registry() {
		registry = new ConcurrentHashMap<>();
	}

	/**
	 * Prints the traversal path from the entry to the root.
	 *
	 * @param path The list representing the path.
	 * @return The string illustrating the path.
	 */
	static String print(List<String> path) {
		StringBuilder sb = new StringBuilder("-");

		for (String step: path) {
			sb.append(" -> ");
			sb.append(step);
		}
		sb.append(" <");

		return sb.toString();
	}

	/**
	 * Adds an entry to the registry.
	 *
	 * @param entry The ID of the entry to add.
	 * @throws DuplicateEntryException Throws DuplicateEntryException if the
	 * entry is already in the registry.
	 */
	void add(String entry) throws DuplicateEntryException {
		if (registry.containsKey(entry)) {
			throw new DuplicateEntryException(String.format(DUPLICATE_ENTRIES,
					entry));
		}
		registry.put(entry, "");
	}

	/**
	 * Adds an entry under a parent entry to the registry.
	 *
	 * @param child The ID of the child entry to add.
	 * @param parent The ID of the parent entry under which to place the child
	 * entry.
	 * @throws DuplicationException Throws this exception if the child entry is
	 * already in the registry.
	 */
	void add(String child, String parent) throws DuplicateEntryException {
		if (registry.containsKey(child)) {
			throw new DuplicateEntryException(String.format(DUPLICATE_ENTRIES,
					child));
		}
		registry.put(child, parent);
	}

	/**
	 * Empties the registry.
	 */
	void clear() {
		registry.clear();
	}

	/**
	 * Exports a snapshot of the registry.
	 *
	 * @return A HashMap typically meant for persistent storage.
	 */
	Map<String, String> export() {
		return new HashMap<String, String>(registry);
	}

	/**
	 * Checks if the entry is stored in the registry.
	 *
	 * @param entry The ID of the entry to check.
	 * @return True if the ID of the entry is present in the registry.
	 */
	boolean has(String entry) {
		return registry.containsKey(entry);
	}

	/**
	 * Checks if there are children IDs under the specified ID.
	 *
	 * @param parentId The ID of the parent to check for.
	 * @return True if there is at least one child ID.
	 */
	boolean hasChild(String parentId) {
		return registry.containsValue(parentId);
	}

	/**
	 * Re-creates the registry with a new hierarchy.
	 *
	 * @param map The map containing the new hierarchy.
	 */
	void importRegistry(Map<String, String> map) {
		registry = new ConcurrentHashMap<String, String>(map);
	}

	/**
	 * Creates a traversal path from the entry to the root.
	 *
	 * @param entry The ID of the entry to start traversing from.
	 * @return A list of entry IDs starting from the entry and ending with the
	 * root.
	 */
	List<String> traverseRoot(String entry) {
		List<String> path = new ArrayList<>();

		if (entry == null) {
			path.add("*");

			return path;
		}

		String eId = entry;

		while (registry.containsKey(eId)) {
			path.add(eId);
			eId = registry.get(eId);
		}
		path.add("*");

		return path;
	}

	/**
	 * Prints a cascading list of entries in this registry.
	 *
	 * @param loader The instance that retrieves other instances.
	 * @param leading The leading space for indented entries.
	 * @param entryId The ID of the entry to start traversing from.
	 * @return The string representing the parent-child relationships between
	 * the entries.
	 */
	String display(AclEntry loader, String leading, String entryId) {
		StringBuilder sb = new StringBuilder();
		List<String> childIds = null;

		if (leading == null) {
			leading = "";
		}
		if (entryId == null) {
			entryId = "";
		}

		childIds = findChildren(entryId);
		for (String childId: childIds) {
			AclEntry entry = loader.retrieveEntry(childId);
			sb.append(leading);
			sb.append("- ");
			sb.append(entry.getEntryDescription());
			sb.append("\n");
			sb.append(this.display(loader, " " + leading, childId));
		}

		return sb.toString();
	}

	/**
	 * Removes an entry from the registry.
	 *
	 * @param entry The ID of the entry to remove from the registry.
	 * @param removeDescendants If true, all child entries and descendants are
	 * removed as well.
	 * @throws NotFoundException Throws this exception if the entry or any of
	 * the descendants (if {@code removeDscendants} is true} are not found.
	 */
	void remove(String entry, boolean removeDescendants)
			throws EntryNotFoundException {
		if (!registry.containsKey(entry)) {
			throw new EntryNotFoundException(String.format(NOT_FOUND, entry));
		}

		if (hasChild(entry)) {
			String parentId = registry.get(entry);
			List<String> childIds = findChildren(entry);

			if (removeDescendants) {
				this.removeDescendants(childIds);
			} else {
				for (String childId: childIds) {
					registry.put(childId, parentId);
				}
			}
		}

		registry.remove(entry);
	}

	/**
	 * Gets the size of the registry.
	 * <p>
	 * Note that the registry contains individual entries, including all parent
	 * and child entries. This means that the size is a simple raw count of all
	 * entries.
	 * </p>
	 *
	 * @return The size of the registry.
	 */
	int size() {
		return registry.size();
	}

	@Override
	public String toString() {
		StringBuilder sb = new StringBuilder();

		for (Map.Entry<String, String> entry: registry.entrySet()) {
			sb.append("\t");
			sb.append(entry.getKey());
			if (entry.getKey().length() >= 8) {
				sb.append("\t - \t");
			} else {
				sb.append("\t\t - \t");
			}
			if (entry.getValue().equals("")) {
				sb.append("*");
			} else {
				sb.append(entry.getValue());
			}
			sb.append("\n");
		}

		return sb.toString();
	}

	private void removeDescendants(List<String> entryIds) {
		for (String entryId: entryIds) {
			String node = entryId;

			registry.remove(node);
			while (hasChild(node)) {
				removeDescendants(findChildren(node));
			}
		}
	}

	private List<String> findChildren(String parentId) {
		List<String> children = new ArrayList<>();

		for (Map.Entry<String, String> entry: registry.entrySet()) {
			if (entry.getValue().equals(parentId)) {
				children.add(entry.getKey());
			}
		}

		return children;
	}
}

File: npm/src/registry.js
/**
 * The registry for resources and roles.
 * @constructor
 */
function Registry() {
  this.registry = {};
}

/**
 * Prints the traversal path from the entry to the root.
 *
 * @param {array} path - The list representing the path.
 */
Registry.display = function (path) {
  var i, out = '-';

  for (i in path) {
    if (path.hasOwnProperty(i)) {
      out += ' -> ';
      out += path[i];
    }
  }
  out += ' <';

  return out;
}

/**
 * Adds an entry to the registry.
 *
 * @param {string} entry - The entry to add.
 * @param {string} [parent] - The parent entry under which to
 * place the child entry.
 * @throws Will throw an error if the entry is already in the
 * registry or if the parent is not in the registry.
 */
Registry.prototype.add = function (entry, parent) {
  if (this.has(entry)) {
    throw new Error(DUPLICATE_ENTRIES.replace(/_entry_/g, entry));
  }
  if (parent) {
    if (!this.has(parent)) {
      throw new Error(ENTRY_NOT_FOUND.replace(/_entry_/g, parent));
    }
    this.registry[entry] = parent;
  } else {
    this.registry[entry] = '';
  }
};

/**
 * Empties the registry.
 */
Registry.prototype.clear = function () {
  this.registry = {};
};

/**
 * Clones the registry to export.
 *
 * @return {object} A clone of the registry.
 */
Registry.prototype.export = function () {
  var i, clone = {};

  for (i in this.registry) {
    if (this.registry.hasOwnProperty(i)) {
      clone[i] = this.registry[i];
    }
  }

  return clone;
};

/**
 * Checks if the entry is stored in the registry.
 *
 * @param {AclEntry} entry - The entry to check.
 * @return {boolean} True if the ID of the entry is present in
 * the registry.
 */
Registry.prototype.has = function (entry) {
  return this.registry[entry] !== undefined;
}

/**
 * Checks if there are children IDs under the specified ID.
 *
 * @param {string} parentId - The ID of the parent to check for.
 * @return {boolean} True if there is at least one child ID.
 */
Registry.prototype.hasChild = function (parentId) {
  for (var i in this.registry) {
    if (this.registry.hasOwnProperty(i)) {
      if (this.registry[i] === parentId) {
        return true;
      }
    }
  }

  return false;
};

/**
 * Re-creates the registry with a new hierarchy.
 *
 * @param {object} map - The map containing the new hierarchy.
 */
Registry.prototype.importRegistry = function (map) {
  var i;

  this.registry = {};
  for (i in map) {
    if (map.hasOwnProperty(i)) {
      this.registry[i] = map[i];
    }
  }
};

/**
 * Creates a traversal path from the entry to the root.
 *
 * @param {AclEntry} entry - The ID of the entry to start
 * traversing from.
 * @return {array} A list of entry IDs starting from the entry and
 * ending with the root.
 */
Registry.prototype.traverseRoot = function (entry) {
  var eId, path = [];

  if (entry == null) {
    path.push('*');

    return path;
  }

  eId = entry;

  while (this.registry[eId] !== undefined) {
    path.push(eId);
    eId = this.registry[eId];
  }
  path.push('*');

  return path;
};

/**
 * Prints a cascading list of entries in this registry.
 *
 * @param {object} loader - An object to retrieve other entries.
 * @param {string} leading - The leading space for indented entries.
 * @param {string} entryId - The ID of the entry to start traversing from.
 * @return {string} The string representing the parent-child
 * relationships between the entries.
 */
Registry.prototype.display = function (loader, leading, entryId) {
  var tis = this,
      childIds, output = [];

  if (!leading) {
      leading = '';
  }
  if (!entryId) {
      entryId = '';
  }

  childIds = findChildren(this.registry, entryId);
  childIds.forEach(function (childId) {
    var entry = loader.retrieveEntry(childId);
    output.push(leading);
    output.push('- ');
    output.push(entry.getEntryDescription());
    output.push('\n');
    output.push(tis.display(loader, ' ' + leading, childId));
  });

  return output.join('');
};

/**
 * Removes an entry from the registry.
 *
 * @param {AclEntry} entry - The entry to remove from the registry.
 * @param {boolean} removeDescendants - If true, all child
 * entries and descendants are removed as well.
 * @throws Will throw an error if the entry or any of the
 * descendants (if 'removeDescendants' is true) are not found.
 */
Registry.prototype.remove = function (entry, removeDescendants) {
  var parentId,
      childIds = [],
      reg = this.registry,
      removed = [];

  if (!this.has(entry)) {
    throw new Error(ENTRY_NOT_FOUND.replace(/_entry_/g, entry));
  }

  if (this.hasChild(entry)) {
    parentId = this.registry[entry];
    childIds = findChildren(this.registry, entry);

    if (removeDescendants) {
      removed = removed.concat(remDescendants(this, childIds));
    } else {
      childIds.forEach(function(childId) {
        reg[childId] = parentId;
      });
    }
  }

  delete this.registry[entry];
  removed.push(entry);

  return removed;
};

Registry.prototype.size = function () {
  return Object.keys(this.registry).length;
};

Registry.prototype.toString = function () {
  var value, key, diff, i,
      len = 0,
      output = [];

  // Get the maximum length
  for (key in this.registry) {
    if (this.registry.hasOwnProperty(key)) {
      if (len < key.length) {
        len = key.length;
      }
    }
  }
  for (key in this.registry) {
    if (this.registry.hasOwnProperty(key)) {
      value = this.registry[key];
      output.push('\t');
      // Add the spaces in front
      diff = len - key.length;
      for (i = 0; i < diff; i++) {
          output.push(' ');
      }
      output.push(key);
      output.push(' - ');
      if (value === '') {
          output.push('*');
      } else {
          output.push(value);
      }
      output.push('\n');
    }
  }

  return output.join('');
};

function remDescendants(reg, entryIds) {
  var removed = [];

  entryIds.forEach(function (entryId) {
    //registry.remove(entryId);
    delete reg.registry[entryId];
    removed.push(entryId);
    while (reg.hasChild(entryId)) {
      removed = removed.concat(remDescendants(reg, findChildren(reg.registry, entryId)));
    }
  });

  return removed;
}

function findChildren(registry, parentId) {
  var key,
      children = [];

  for (key in registry) {
    if (registry.hasOwnProperty(key)) {
      if (registry[key] === parentId) {
        children.push(key);
      }
    }
  }

  return children;
}

module.exports = Registry;
